File: sdk-ts/src/crypto.ts
import { randomBytes } from 'crypto';
import * as secp from '@noble/secp256k1';
import { sha256 } from '@noble/hashes/sha256';
import { hkdf } from '@noble/hashes/hkdf';
import { gcm } from '@noble/ciphers/aes';
import { randomBytes as nobleRandom } from '@noble/hashes/utils';

/**
 * Generate a secure random private key
 */
export function generatePrivateKey(): Uint8Array {
  return secp.utils.randomPrivateKey();
}

/**
 * Get public key from private key
 */
export function getPublicKey(privateKey: Uint8Array): Uint8Array {
  return secp.getPublicKey(privateKey);
}

/**
 * Sign a message with private key
 */
export async function signMessage(
  message: Uint8Array,
  privateKey: Uint8Array
): Promise<Uint8Array> {
  const hash = sha256(message);
  const signature = await secp.signAsync(hash, privateKey);
  return signature;
}

/**
 * Verify a signature
 */
export async function verifySignature(
  message: Uint8Array,
  signature: Uint8Array,
  publicKey: Uint8Array
): Promise<boolean> {
  const hash = sha256(message);
  return secp.verify(signature, hash, publicKey);
}

/**
 * Derive shared secret using ECDH
 */
export function deriveSharedSecret(
  privateKey: Uint8Array,
  publicKey: Uint8Array
): Uint8Array {
  const shared = secp.getSharedSecret(privateKey, publicKey);
  // Remove the 0x04 prefix if present (uncompressed key marker)
  return shared.slice(shared.length === 33 ? 1 : 0);
}

/**
 * Derive session key from shared secret using HKDF
 */
export function deriveSessionKey(sharedSecret: Uint8Array): Uint8Array {
  const salt = new Uint8Array(32); // Zero salt
  const info = new TextEncoder().encode('H3DAC');
  const sessionKey = hkdf(sha256, sharedSecret, salt, info, 32);
  return sessionKey;
}

/**
 * Encrypt payload with AES-GCM
 */
export function encryptPayload(
  payload: Uint8Array,
  sessionKey: Uint8Array
): { encrypted: Uint8Array; nonce: Uint8Array } {
  const nonce = nobleRandom(12); // 96-bit nonce for GCM
  const cipher = gcm(sessionKey, nonce);
  const encrypted = cipher.encrypt(payload);
  return { encrypted, nonce };
}

/**
 * Decrypt payload with AES-GCM
 */
export function decryptPayload(
  encrypted: Uint8Array,
  sessionKey: Uint8Array,
  nonce: Uint8Array
): Uint8Array {
  const cipher = gcm(sessionKey, nonce);
  return cipher.decrypt(encrypted);
}

/**
 * Generate a random nonce
 */
export function generateNonce(): string {
  return randomBytes(32).toString('hex');
}

/**
 * Hash data with SHA256
 */
export function hashData(data: Uint8Array): Uint8Array {
  return sha256(data);
}

/**
 * Concatenate multiple byte arrays
 */
export function concatBytes(...arrays: Uint8Array[]): Uint8Array {
  const totalLength = arrays.reduce((acc, arr) => acc + arr.length, 0);
  const result = new Uint8Array(totalLength);
  let offset = 0;
  for (const arr of arrays) {
    result.set(arr, offset);
    offset += arr.length;
  }
  return result;
}

/**
 * Convert hex string to Uint8Array
 */
export function hexToBytes(hex: string): Uint8Array {
  if (hex.startsWith('0x')) hex = hex.slice(2);
  return new Uint8Array(
    hex.match(/.{1,2}/g)?.map((byte) => parseInt(byte, 16)) || []
  );
}

/**
 * Convert Uint8Array to hex string
 */
export function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('');
}

File: sdk-rust/src/crypto.rs
use secp256k1::{ecdh::SharedSecret, PublicKey, Secp256k1, SecretKey};
use sha2::{Digest, Sha256};
use hkdf::Hkdf;
use aes_gcm::{
    aead::{Aead, KeyInit, OsRng},
    Aes256Gcm, Nonce,
};
use rand::RngCore;

use crate::error::{H3DACError, Result};

/// Generate a new random private key
pub fn generate_private_key() -> SecretKey {
    let secp = Secp256k1::new();
    SecretKey::new(&mut rand::thread_rng())
}

/// Get public key from private key
pub fn get_public_key(private_key: &SecretKey) -> PublicKey {
    let secp = Secp256k1::new();
    PublicKey::from_secret_key(&secp, private_key)
}

/// Sign a message with private key
pub fn sign_message(message: &[u8], private_key: &SecretKey) -> Result<Vec<u8>> {
    let secp = Secp256k1::new();
    let hash = hash_data(message);
    let message = secp256k1::Message::from_digest_slice(&hash)
        .map_err(|e| H3DACError::CryptoError(format!("Invalid message: {}", e)))?;
    
    let signature = secp.sign_ecdsa(&message, private_key);
    Ok(signature.serialize_compact().to_vec())
}

/// Verify a signature
pub fn verify_signature(
    message: &[u8],
    signature: &[u8],
    public_key: &PublicKey,
) -> Result<bool> {
    let secp = Secp256k1::new();
    let hash = hash_data(message);
    let message = secp256k1::Message::from_digest_slice(&hash)
        .map_err(|e| H3DACError::CryptoError(format!("Invalid message: {}", e)))?;
    
    let sig = secp256k1::ecdsa::Signature::from_compact(signature)
        .map_err(|e| H3DACError::CryptoError(format!("Invalid signature: {}", e)))?;
    
    Ok(secp.verify_ecdsa(&message, &sig, public_key).is_ok())
}

/// Derive shared secret using ECDH
pub fn derive_shared_secret(
    private_key: &SecretKey,
    public_key: &PublicKey,
) -> Result<Vec<u8>> {
    let shared = SharedSecret::new(public_key, private_key);
    Ok(shared.secret_bytes().to_vec())
}

/// Derive session key from shared secret using HKDF
pub fn derive_session_key(shared_secret: &[u8]) -> Result<Vec<u8>> {
    let hk = Hkdf::<Sha256>::new(None, shared_secret);
    let mut session_key = vec![0u8; 32];
    hk.expand(b"H3DAC", &mut session_key)
        .map_err(|e| H3DACError::CryptoError(format!("HKDF failed: {}", e)))?;
    Ok(session_key)
}

/// Encrypt payload with AES-GCM
pub fn encrypt_payload(payload: &[u8], session_key: &[u8]) -> Result<(Vec<u8>, Vec<u8>)> {
    if session_key.len() != 32 {
        return Err(H3DACError::CryptoError(
            "Session key must be 32 bytes".to_string(),
        ));
    }

    let key = aes_gcm::Key::<Aes256Gcm>::from_slice(session_key);
    let cipher = Aes256Gcm::new(key);

    let mut nonce_bytes = [0u8; 12];
    rand::thread_rng().fill_bytes(&mut nonce_bytes);
    let nonce = Nonce::from_slice(&nonce_bytes);

    let encrypted = cipher
        .encrypt(nonce, payload)
        .map_err(|e| H3DACError::CryptoError(format!("Encryption failed: {}", e)))?;

    Ok((encrypted, nonce_bytes.to_vec()))
}

/// Decrypt payload with AES-GCM
pub fn decrypt_payload(
    encrypted: &[u8],
    session_key: &[u8],
    nonce: &[u8],
) -> Result<Vec<u8>> {
    if session_key.len() != 32 {
        return Err(H3DACError::CryptoError(
            "Session key must be 32 bytes".to_string(),
        ));
    }

    if nonce.len() != 12 {
        return Err(H3DACError::CryptoError(
            "Nonce must be 12 bytes".to_string(),
        ));
    }

    let key = aes_gcm::Key::<Aes256Gcm>::from_slice(session_key);
    let cipher = Aes256Gcm::new(key);
    let nonce = Nonce::from_slice(nonce);

    let decrypted = cipher
        .decrypt(nonce, encrypted)
        .map_err(|e| H3DACError::CryptoError(format!("Decryption failed: {}", e)))?;

    Ok(decrypted)
}

/// Generate a random nonce
pub fn generate_nonce() -> String {
    let mut nonce = vec![0u8; 32];
    rand::thread_rng().fill_bytes(&mut nonce);
    hex::encode(nonce)
}

/// Hash data with SHA256
pub fn hash_data(data: &[u8]) -> Vec<u8> {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hasher.finalize().to_vec()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_key_generation() {
        let private_key = generate_private_key();
        let public_key = get_public_key(&private_key);
        assert_eq!(public_key.serialize().len(), 33);
    }

    #[test]
    fn test_sign_verify() {
        let private_key = generate_private_key();
        let public_key = get_public_key(&private_key);
        let message = b"test message";

        let signature = sign_message(message, &private_key).unwrap();
        let valid = verify_signature(message, &signature, &public_key).unwrap();
        assert!(valid);
    }

    #[test]
    fn test_shared_secret() {
        let private1 = generate_private_key();
        let public1 = get_public_key(&private1);

        let private2 = generate_private_key();
        let public2 = get_public_key(&private2);

        let shared1 = derive_shared_secret(&private1, &public2).unwrap();
        let shared2 = derive_shared_secret(&private2, &public1).unwrap();

        assert_eq!(shared1, shared2);
    }

    #[test]
    fn test_encryption() {
        let session_key = vec![0u8; 32];
        let payload = b"secret message";

        let (encrypted, nonce) = encrypt_payload(payload, &session_key).unwrap();
        let decrypted = decrypt_payload(&encrypted, &session_key, &nonce).unwrap();

        assert_eq!(payload.to_vec(), decrypted);
    }
}
